python/demo/demo_pml/demo_pml.py

```diff
@@ -0,0 +1,606 @@
+# \end{align}
+#
+# Now we define some other problem specific parameters:
+
+wl0 = 0.4  # Wavelength of the background field
+n_bkg = 1  # Background refractive index
+eps_bkg = n_bkg**2  # Background relative permittivity
+k0 = 2 * np.pi / wl0  # Wavevector of the background field
+theta = 0  # Angle of incidence of the background field
+
+# We use a degree 3
+# [Nedelec (first kind)](https://defelement.com/elements/nedelec1.html)
+# element to represent the electric field:
+#
+
+degree = 3
+curl_el = ufl.FiniteElement("N1curl", domain.ufl_cell(), degree)
+V = fem.FunctionSpace(domain, curl_el)
+
+# Next, we interpolate $\mathbf{E}_b$ into the function space $V$, define our
+# trial and test function, and the integration domains:
+
+# +
+Eb = fem.Function(V)
+f = partial(background_field, theta, n_bkg, k0)
+Eb.interpolate(f)
+
+# Definition of Trial and Test functions
+Es = ufl.TrialFunction(V)
+v = ufl.TestFunction(V)
+
+# Definition of 3d fields
+Es_3d = ufl.as_vector((Es[0], Es[1], 0))
+v_3d = ufl.as_vector((v[0], v[1], 0))
+
+# Measures for subdomains
+dx = ufl.Measure("dx", domain, subdomain_data=cell_tags)
+dDom = dx((au_tag, bkg_tag))
+dPml_xy = dx(pml_tag)
+dPml_x = dx(pml_tag + 1)
+dPml_y = dx(pml_tag + 2)
+# -
+
+# Let's now define the relative permittivity $\varepsilon_m$
+# of the gold wire at $400nm$ (data taken from
+# [*Olmon et al. 2012*](https://doi.org/10.1103/PhysRevB.86.235147)
+# , and for a quick reference have a look at [refractiveindex.info](
+# https://refractiveindex.info/?shelf=main&book=Au&page=Olmon-sc
+# )):
+
+# Definition of relative permittivity for Au @400nm
+eps_au = -1.0782 + 1j * 5.8089
+
+
+# We can now define a space function for the permittivity
+# $\varepsilon$ that takes the value $\varepsilon_m$
+# for cells inside the wire, while it takes the value of the
+# background permittivity $\varepsilon_b$ in the background region:
+
+D = fem.FunctionSpace(domain, ("DG", 0))
+eps = fem.Function(D)
+au_cells = cell_tags.find(au_tag)
+bkg_cells = cell_tags.find(bkg_tag)
+eps.x.array[au_cells] = np.full_like(
+    au_cells, eps_au, dtype=np.complex128)
+eps.x.array[bkg_cells] = np.full_like(bkg_cells, eps_bkg, dtype=np.complex128)
+eps.x.scatter_forward()
+
+# Now we need to define our weak form in DOLFINx.
+# Let's write the PML weak form first. As a first step,
+# we can define our new complex coordinates as:
+
+# +
+x = ufl.SpatialCoordinate(domain)
+alpha = 1
+
+# PML corners
+xy_pml = ufl.as_vector((pml_coordinates(x[0], alpha, k0, l_dom, l_pml),
+                        pml_coordinates(x[1], alpha, k0, l_dom, l_pml)))
+
+# PML rectangles along x
+x_pml = ufl.as_vector((pml_coordinates(x[0], alpha, k0, l_dom, l_pml), x[1]))
+
+# PML rectangles along y
+y_pml = ufl.as_vector((x[0], pml_coordinates(x[1], alpha, k0, l_dom, l_pml)))
+# -
+
+# We can then express this coordinate systems as
+# a material transformation within the PML region. In other words,
+# the PML region can be interpreted as a material
```

Comment:
This is known as a UPML (uniaxial PML). Maybe this could be mentioned here somewhere.